File: solutions/python/twelve-days/1/twelve_days.py
GIFTS = [
    "a Partridge in a Pear Tree",
    "two Turtle Doves",
    "three French Hens",
    "four Calling Birds",
    "five Gold Rings",
    "six Geese-a-Laying",
    "seven Swans-a-Swimming",
    "eight Maids-a-Milking",
    "nine Ladies Dancing",
    "ten Lords-a-Leaping",
    "eleven Pipers Piping",
    "twelve Drummers Drumming",
]

ORDINALS = ["first", "second", "third"] + [
    word + "th"
    for word in [
        "four",
        "fif",
        "six",
        "seven",
        "eigh",
        "nin",
        "ten",
        "eleven",
        "twelf",
    ]
]


def oxford_join(parts: list[str]) -> str:
    if len(parts) == 1:
        return parts[0]
    return ", ".join(parts[:-1] + ["and " + parts[-1]])


VERSES = [
    f"On the {ORDINALS[day - 1]} day of Christmas my true love gave to me: "
    f"{oxford_join(GIFTS[day - 1 : : -1])}."
    for day in range(1, 13)
]


def recite(start_verse: int, end_verse: int) -> list[str]:
    return VERSES[start_verse - 1 : end_verse]

File: solutions/go/twelve-days/1/twelve_days.go
package twelve

import (
	"fmt"
	"slices"
	"strings"
)

var gifts = []string{
	"a Partridge in a Pear Tree",
	"two Turtle Doves",
	"three French Hens",
	"four Calling Birds",
	"five Gold Rings",
	"six Geese-a-Laying",
	"seven Swans-a-Swimming",
	"eight Maids-a-Milking",
	"nine Ladies Dancing",
	"ten Lords-a-Leaping",
	"eleven Pipers Piping",
	"twelve Drummers Drumming",
}

var ordinals = []string{"first", "second", "third",
	"fourth",
	"fifth",
	"sixth",
	"seventh",
	"eighth",
	"ninth",
	"tenth",
	"eleventh",
	"twelfth",
}

func oxfordJoin(parts []string) string {
	if len(parts) == 0 {
		return ""
	}
	if len(parts) == 1 {
		return parts[0]
	}
	return strings.Join(
		append(parts[:len(parts)-1], "and "+parts[len(parts)-1]),
		", ",
	)
}

func Verse(i int) string {
	myGifts := slices.Clone(gifts[:i])
	slices.Reverse(myGifts)
	return fmt.Sprintf(
		"On the %s day of Christmas my true love gave to me: %s.",
		ordinals[i-1], oxfordJoin(myGifts),
	)
}

func Song() string {
	verses := []string{}
	for i := 1; i <= 12; i++ {
		verses = append(verses, Verse(i))
	}
	return strings.Join(verses, "\n")
}
